### ISSUE TITLE
"DONE" too soon

### ISSUE THREAD

**Issue by solardiz** [magnumripper/JohnTheRipper#3825]
In the example below, we have the final status line saying "DONE" and giving a non-final guess count (10k instead of 50k) before many more successful cracks are produced. Previously, I thought we were getting "DONE" when the full last buffer is tested against some salt, but not yet against all salts. But this time it's with just 1 salt total. So apparently we're getting "DONE" when we've merely buffered the last batch of candidates, even when running against 1 salt. This issue is also seen with other cracking modes - it is not specific to the increasing mask lengths (as is seen in this example). Perhaps we should look into fixing it because it's bad user experience to see "DONE" prematurely and because runs like the below end up never showing the real final status line with the correct guess count, etc.

```
$ rm -f john.pot john.log john.rec
$ ./john -mask='?d' -min-len=1 -max-len=5 -verb=1 pw-bcrypt-50k 
ZTEX 2 bus:2 dev:65 Frequency:150 150 150 150 
ZTEX 3 bus:2 dev:66 Frequency:150 150 150 150 
ZTEX 1 bus:2 dev:67 Frequency:150 150 150 150 
ZTEX 4 bus:2 dev:68 Frequency:150 150 150 150 
Warning: detected hash type "bcrypt-ztex", but the string is also recognized as "bcrypt"
Use the "--format=bcrypt" option to force loading these as that type instead
Using default input encoding: UTF-8
Loaded 50000 password hashes with no different salts (bcrypt-ztex [Blowfish ZTEX])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:00  (1) 0g/s 0p/s 0c/s 0C/s
Warning: salt with 50000 hashes, onboard comparators support up to 512 hashes/salt, turned off
10g 0:00:00:00 9.09% (2) (ETA: 16:08:03) 90.90g/s 90.90p/s 90.90c/s 4545KC/s 11..17
100g 0:00:00:00 9.91% (3) (ETA: 16:08:03) 769.2g/s 846.1p/s 846.1c/s 42300KC/s 111..177
1000g 0:00:00:00 9.99% (4) (ETA: 16:08:03) 5263g/s 5842p/s 5842c/s 291573KC/s 1111..1777
10000g 0:00:00:00 10.99% (5) (ETA: 16:08:04) 13513g/s 15013p/s 15013c/s 737025KC/s 7771..7779
10000g 0:00:00:00 DONE (5) (2019-04-15 16:08) 13513g/s 15013p/s 15013c/s 737025KC/s 7771..7779
Session completed
$ wc -l john.pot
50000 john.pot
```

**Comment by solardiz** [comment]
Looks like things got worse with respect to this issue fairly recently. I recall this test printing 239g on its final line a few months ago:

```
$ ./john -mask='?l?l?l?l' -format=bcrypt-ztex -verb=1 pw-fake-len4
Using default input encoding: UTF-8
Loaded 239 password hashes with 239 different salts (bcrypt-ztex [Blowfish ZTEX])
4 device(s) ZTEX 1.15y ready
Press 'q' or Ctrl-C to abort, almost any other key for status
2g 0:00:00:02  0.8032g/s 0p/s 451752c/s 451752C/s aaaa..aaba
4g 0:00:00:04  0.9345g/s 0p/s 459932c/s 459932C/s aaaa..aaba
6g 0:00:00:05  1.096g/s 0p/s 462695c/s 462695C/s neil..aaba
9g 0:00:00:09  0.9336g/s 0p/s 466748c/s 466748C/s aaaa..aaba
17g 0:00:00:16  1.049g/s 0p/s 468693c/s 468693C/s lynn..aaba
21g 0:00:00:18  1.130g/s 0p/s 469197c/s 469197C/s duke..aaba
24g 0:00:00:20  1.178g/s 0p/s 469614c/s 469614C/s chat..aaba
26g 0:00:00:21  1.206g/s 0p/s 469780c/s 469780C/s coke..aaba
86g 0:00:01:01  1.398g/s 0p/s 470979c/s 470979C/s beta..aaba
88g 0:00:01:03  1.390g/s 0p/s 470989c/s 470989C/s lima..aaba
91g 0:00:01:05  1.385g/s 0p/s 471048c/s 471048C/s aaaa..aaba
95g 0:00:01:08  1.383g/s 0p/s 471082c/s 471082C/s aaaa..aaba
110g 0:00:01:20  1.365g/s 0p/s 471253c/s 471253C/s aaaa..aaba
194g 0:00:02:22 61.54% (ETA: 14:02:45) 1.360g/s 1972p/s 471322c/s 471322C/s kali..aamk
Session completed
$ wc -l john.pot 
239 john.pot
```

**Comment by solardiz** [comment]
For those reading this via e-mail - I've made major edits to my previous comment on "DONE too soon".

**Comment by solardiz** [comment]
I've just moved this issue to the 1.9.0-jumbo-1 milestone since we appear to have a recent regression here. I'd like us to try and find and fix that regression before the release. Fixing the rest of this issue is probably a lot more work and can wait for longer.

**Comment by solardiz** [comment]
Per my testing, the regression appeared between March 29 and April 15. I can't find a commit description referencing anything relevant. I suspect it's some side-effect of a bcrypt-ztex change, but need to confirm that. Reverting the clock rate from 150 to 141 doesn't make a difference.

**Comment by Apingis** [comment]
Except for what's written in the headline, you're describing the effect described in #3824.

**Comment by solardiz** [comment]
Thanks @Apingis. I'm not yet sure if the issue I reported here is entirely or only partially a duplicate of #3824. I hope Claudio or someone else will fix #3824, and then I'll retest what's in my first comment here and close this issue if what I reported there is gone.

**Comment by solardiz** [comment]
The final line does show the full guess count in my testing now, so we've fixed that aspect of the issue. I'll close this for now, and we may open another issue for not starting to say "DONE" too soon later.
